File: programming/mainstream/python/python3/Python3/Python3LexerBase.py
from typing import TextIO
from antlr4 import *
from antlr4.Token import CommonToken
from Python3Parser import Python3Parser;

import sys
from typing import TextIO
import re


class Python3LexerBase(Lexer):
    NEW_LINE_PATTERN = re.compile('[^\r\n\f]+')
    SPACES_PATTERN = re.compile('[\r\n\f]+')

    def __init__(self, input: InputStream, output: TextIO = sys.stdout):
        super().__init__(input, output)
        self.tokens = []
        self.indents = []
        self.opened = 0
    
    def reset(self):
        self.tokens = []
        self.indents = []
        self.opened = 0
        super().reset()
    
    def emitToken(self, token):
        self._token = token
        self.tokens.append(token)
    
    def nextToken(self):
        # Check if the end-of-file is ahead and there are still some DEDENTS expected.
        if self._input.LA(1) == Python3Parser.EOF and len(self.indents) != 0:
            # Remove any trailing EOF tokens from our buffer.
            self.tokens = [token for token in self.tokens if token.type != Python3Parser.EOF]

            # First emit an extra line break that serves as the end of the statement.
            self.emitToken(self.commonToken(Python3Parser.NEWLINE, '\n'))

            # Now emit as much DEDENT tokens as needed.
            while len(self.indents) != 0:
                self.emitToken(self.createDedent())
                self.indents.pop()

            # Put the EOF back on the token stream.
            self.emitToken(self.commonToken(Python3Parser.EOF, '<EOF>'))

        next_ = super().nextToken()
        return next_ if len(self.tokens) == 0 else self.tokens.pop(0)

    def createDedent(self):
        return self.commonToken(Python3Parser.DEDENT, '')

    def commonToken(self, type_: int, text: str):
        stop = self.getCharIndex() - 1
        start = stop if text == '' else stop - len(text) + 1
        return CommonToken(self._tokenFactorySourcePair, type_, Lexer.DEFAULT_TOKEN_CHANNEL, start, stop)

    def getIndentationCount(self, whitespace: str):
        count = 0
        for c in whitespace:
            if c == '\t':
                count += 8 - count % 8
            else:
                count += 1
        return count

    def atStartOfInput(self):
        return self.getCharIndex() == 0

    def openBrace(self):
        self.opened += 1

    def closeBrace(self):
        self.opened -= 1

    def onNewLine(self):
        new_line = self.NEW_LINE_PATTERN.sub('', self.text)
        spaces = self.SPACES_PATTERN.sub('', self.text)

        # Strip newlines inside open clauses except if we are near EOF. We keep NEWLINEs near EOF to
        # satisfy the final newline needed by the single_put rule used by the REPL.
        next_ = self._input.LA(1)
        next_next = self._input.LA(2)

        if self.opened > 0 or (next_next != -1 and next_ in (10, 13, 35)):
            self.skip()
        else:
            self.emitToken(self.commonToken(Python3Parser.NEWLINE, new_line))
            indent = self.getIndentationCount(spaces)
            previous = 0 if len(self.indents) == 0 else self.indents[-1]

            if indent == previous:
                self.skip()
            elif indent > previous:
                self.indents.append(indent)
                self.emitToken(self.commonToken(Python3Parser.INDENT, spaces))
            else:
                while len(self.indents) > 0 and self.indents[-1] > indent:
                    self.emitToken(self.createDedent())
                    self.indents.pop()

File: python/python3/Go/python3_lexer_base.go
package parser

import (
	"regexp"

	"github.com/antlr4-go/antlr/v4"
)

type Python3LexerBase struct {
	*antlr.BaseLexer
	tokens    []antlr.Token
	indents   []int
	Opened    int
	lastToken antlr.Token
}

func (l *Python3LexerBase) Emit() antlr.Token {
	t := l.BaseLexer.Emit()
	return t
}

func (l *Python3LexerBase) EmitToken(t antlr.Token) {
	l.BaseLexer.EmitToken(t)
	l.tokens = append(l.tokens, t)
}

func (l *Python3LexerBase) MakeCommonToken(ttype int, text string) antlr.Token {
	stop := l.TokenStartCharIndex - 1
	start := stop
	if len(text) != 0 {
		start = stop - len(text) + 1
	}
	ctf := l.GetTokenFactory()
	t := ctf.Create(
		l.GetTokenSourceCharStreamPair(),
		ttype,
		text,
		antlr.TokenDefaultChannel,
		start,
		l.TokenStartCharIndex-1,
		l.TokenStartLine,
		l.TokenStartColumn)
	return t
}

func (l *Python3LexerBase) CreateDedent() antlr.Token {
	dedent := l.MakeCommonToken(Python3ParserDEDENT, "")
	//dedent.Line = LastToken.GetLine()
	return dedent
}

func (l *Python3LexerBase) NextToken() antlr.Token {
	if l.GetInputStream().LA(1) == antlr.TokenEOF && len(l.indents) != 0 {
		var filtered []antlr.Token
		for _, value := range l.tokens {
			if value.GetTokenType() != antlr.TokenEOF {
				filtered = append(filtered, value)
			}
		}
		l.tokens = filtered
		l.EmitToken(l.MakeCommonToken(Python3ParserNEWLINE, "\n"))
		for len(l.indents) != 0 {
			l.EmitToken(l.CreateDedent())
			l.indents = l.indents[:len(l.indents)-1]
		}
		l.EmitToken(l.MakeCommonToken(antlr.TokenEOF, "<EOF>"))
	}

	next := l.BaseLexer.NextToken()
	if next.GetChannel() == antlr.TokenDefaultChannel {
		l.lastToken = next
	}
	if len(l.tokens) == 0 {
		return next
	} else {
		x := l.tokens[0]
		l.tokens = l.tokens[1:]
		return x
	}
}

func (l *Python3LexerBase) GetIndentationCount(spaces string) int {
	count := 0
	for _, ch := range spaces {
		if ch == '\t' {
			count += 8 - (count % 8)
		} else {
			count += 1
		}
	}
	return count
}

func (l *Python3LexerBase) atStartOfInput() bool {
	return l.TokenStartColumn == 0 && l.TokenStartLine == 1
}

func (l *Python3LexerBase) openBrace() {
	l.Opened++
}

func (l *Python3LexerBase) closeBrace() {
	l.Opened--
}

func (l *Python3LexerBase) onNewLine() {
	newLineRegex := regexp.MustCompile(`[^\r\n\f]+`)
	newLine := newLineRegex.ReplaceAllString(l.GetText(), "")
	spacesRegex := regexp.MustCompile(`[\r\n\f]+`)
	spaces := spacesRegex.ReplaceAllString(l.GetText(), "")
	next := l.GetInputStream().LA(1)
	nextnext := l.GetInputStream().LA(2)
	if l.Opened > 0 || (nextnext != -1 && (next == '\r' || next == '\n' || next == '\f' || next == '#')) {
		l.Skip()
	} else {
		l.EmitToken(l.MakeCommonToken(Python3LexerNEWLINE, newLine))
		indent := l.GetIndentationCount(spaces)
		previous := 0
		if len(l.indents) == 0 {
			previous = 0
		} else {
			previous = l.indents[len(l.indents)-1]
		}
		if indent == previous {
			l.Skip()
		} else if indent > previous {
			l.indents = append(l.indents, indent)
			l.EmitToken(l.MakeCommonToken(Python3ParserINDENT, spaces))
		} else {
			for len(l.indents) != 0 && l.indents[len(l.indents)-1] > indent {
				l.EmitToken(l.CreateDedent())
				l.indents = l.indents[:len(l.indents)-1]
			}
		}
	}
}

func (l *Python3LexerBase) Reset() {
	l.tokens = make([]antlr.Token, 0)
	l.indents = make([]int, 0)
	l.Opened = 0
	//l.lastToken = null
	l.BaseLexer.Reset()
}
